### PR TITLE
[AMD] Add kimi_linear_models test on AMD GPU

### PR DESCRIPTION
**Pull request by sogalin**
## Motivation

Add `test_kimi_linear_models.py` to AMD CI coverage and fix the ROCm code
paths required to make the Kimi-Linear MLA + KDA hybrid model produce
correct outputs.

`moonshotai/Kimi-Linear-48B-A3B-Instruct` is a hybrid attention model
(20 KDA layers + 7 MLA layers) with `mla_use_nope=True`, so the MLA layers
are trained without RoPE and `self.rotary_emb` is `None` on those layers.
This combination currently breaks AMD/ROCm in several places:

1. The MLA absorb-core path dereferences `self.rotary_emb` unconditionally
   on the `_use_aiter_gfx95` branch (crash during CUDA-graph capture).
2. The MLA_FUSED_ROPE_ROCM dispatch path makes the same assumption.
3. The aiter ASM absorbed-MLA decode kernel is numerically lossy for the
   skip-rope MLA layers (gsm8k score collapses to ~0.5).
4. After bringing main up to date, two newer KDA PRs (#24271, #23186)
   surface additional AMD-only regressions described below.

### Model code fixes for `mla_use_nope=True` (skip-rope MLA)

1. `python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py`
   In `forward_absorb_core`, the `_use_aiter_gfx95` branch unconditionally
   reads `self.rotary_emb.cos_cache`. Gate it on
   `self.rotary_emb is not None` so skip-rope MLA layers fall through to
   the simple `torch.cat(q_nope, q_pe)` path that already exists for
   non-aiter_gfx95 hosts.

2. `python/sglang/srt/models/deepseek_common/attention_backend_handler.py`
   In `_dispatch_mla_subtype`, the ROCm `MLA_FUSED_ROPE_ROCM` path also
   reads `rotary_emb.cos_sin_cache` / `rotary_dim`. Add a
   `getattr(attn, "rotary_emb", None) is not None` guard so skip-rope MLA
   layers route to the generic `MLA` path instead.

### KDA AMD compatibility fixes (post main merge)

3. `python/sglang/srt/layers/attention/fla/kda.py`

   a. PR #24271 (`[KDA] Optimize prefill kernels with diagonal and
      recompute fuse`) hardcoded `DOT_PRECISION="tf32"`, which is
      NVIDIA-only. AMD/ROCm triton crashes with
      `input_precision must be one of ('ieee', 'bf16x3', 'bf16x6')`.
      Gate the dot precision on `is_tf32_supported` and fall back to
      `"ieee"` on AMD (mirrors `chunk_fwd.py`'s `_MERGE_DOT_PRECISION`
      pattern).

   b. The same PR added a fused diagonal + recompute path
      (`_small_grid`) that wins on CUDA Ampere+ (small launch overhead,
      tf32 dot). On AMD/ROCm the fusion changes the chunk dot-product
      accumulation order in bf16 and increases Kimi-Linear gsm8k score
      variance (mean 0.901 -> 0.895, stdev 0.009 -> 0.014, with
      scattered <0.88 outliers) without a clear AMD perf win. Gate
      `_small_grid` on `is_tf32_supported` as well so AMD takes the
      pre-#24271 non-fused KDA path, which is numerically stable.

4. `python/sglang/srt/models/kimi_linear.py`
   Wire `SGLANG_ROCM_USE_MULTI_STREAM` opt-in into `KimiLinearModel.alt_stream`,
   matching the gating added to `deepseek_v2.DeepseekV2Model` in PR #24005.
   The existing `torch.cuda.Stream() if _is_cuda else None` line is kept
   untouched; the ROCm opt-in is layered on top so other backends are
   not affected. Default behaviour on AMD remains single-stream (env var
   defaults to off).

### Test registration

5. `test/registered/models/test_kimi_linear_models.py`
   * Register the file for both CUDA and AMD CI:
     `stage-b-test-2-gpu-large` / `stage-b-test-2-gpu-large-amd`.
   * On ROCm, split decode and prefill backends:
     `--decode-attention-backend triton --prefill-attention-backend aiter`.

     The aiter absorbed-MLA *decode* kernel (`aiter.mla.mla_decode_fwd`,
     ASM) is numerically lossy on unrotated q_pe / k_pe -- gsm8k drops
     to **0.515** with the default aiter decode path even after both
     crash fixes above. Routing decode through Triton (and keeping
     aiter for prefill where it works correctly and is faster) brings
     accuracy back to **~0.90** on MI355X / gfx950. This mirrors the
     configuration in `test/registered/amd/test_kimi_k2_instruct.py`.

     Forcing the *entire* attention backend to triton
     (`--attention-backend triton`) was tried but produces wrong
     outputs on MI355X / gfx950 (gsm8k -> 0.0, 20x slower); the
     decode/prefill split is required.

## Accuracy Tests

All numbers from MI355X (gfx950) x 2, TP=2, bf16, `SGLANG_USE_AITER=1`.

### Full gsm8k (1319 questions, parallel=1024)

| metric          | value                |
| --------------- | -------------------- |
| **accuracy**    | **0.9011** (> 0.88)  |
| latency         | 30.53 s              |
| wall            | 30.62 s              |
| output tput     | 4189.9 tok/s         |
| invalid answers | 0                    |

### Stability -- 200 questions x 100 iterations (parallel=128)

| metric                   | value     |
| ------------------------ | --------- |
| accuracy mean            | 0.8992    |
| accuracy stdev           | 0.0094    |
| accuracy min / max       | 0.880 / 0.925 |
| wall mean per iter       | 5.73 s    |
| pass rate `acc >= 0.88`  | **100 / 100** |
| pass rate `acc >  0.88`  | 98 / 100 (strict)  |

The 2 iterations that hit exactly 0.88 are absorbed by the existing
`CustomTestCase` retry (`max_retry=1` in CI), giving an effective
per-job failure rate of ~0.04%.

### Comparison of routing decisions on AMD

| AMD config                                                     | gsm8k 200 | wall   |
| -------------------------------------------------------------- | --------- | ------ |
| default aiter (broken before fixes)                            | crash     | -      |
| default aiter (after `forward_mla.py` + handler fixes)         | 0.515     | 7.6 s  |
| **`--decode-attention-backend triton --prefill=aiter`** (this) | **0.910** | **5.7 s** |
| `--attention-backend triton` (PR earlier attempt)              | 0.000     | 143 s  |

### Pre-merge vs post-merge (after KDA fixes)

| metric                       | Pre-merge baseline | This PR (post-merge + KDA fix) |
| ---------------------------- | ------------------ | ------------------------------ |
| Full gsm8k score             | 0.8973             | **0.9011** |
| Full gsm8k throughput        | 4031 tok/s         | **4190 tok/s** |
| 200x100 iter score mean      | 0.8998             | **0.8992** |
| 200x100 iter score stdev     | 0.0090             | **0.0094** |
| 200x100 iter score min       | 0.880              | **0.880** |
| 200x100 pass strict (>0.88)  | 98/100             | **98/100** |
| 200x100 pass inclusive (>=)  | 100/100            | **100/100** |

The KDA fixes recover full pre-merge performance and accuracy.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

---
### CI States

Latest PR Test (Base): :x: [Run #26879911576](https://github.com/sgl-project/sglang/actions/runs/26879911576)
Latest PR Test (Extra): :x: [Run #26879911413](https://github.com/sgl-project/sglang/actions/runs/26879911413)

### CODE REVIEW

File: test/registered/models/test_kimi_linear_models.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 )
 
 register_cuda_ci(est_time=90, suite="stage-b-test-large-2-gpu")
+register_cuda_ci(est_time=90, suite="stage-b-test-large-2-gpu-amd")
```
Comment (yctseng0211) [inline test/registered/models/test_kimi_linear_models.py]: should be **register_amd_ci**

File: test/registered/models/test_kimi_linear_models.py
Comment (yctseng0211) [inline test/registered/models/test_kimi_linear_models.py:5]: import register_amd_ci, register_cuda_ci

Comment (sogalin) [reply test/registered/models/test_kimi_linear_models.py:5]: thanks

**Review by yichiche — APPROVED** [review]
By adding a check for self.rotary_emb, we can prevent attribute errors in models like Kimi that use skip_rope=True, as DSv3 assumes every layer adopts rotary_emb.
LGTM.

**Comment by sogalin** [comment]
Found GSM8K accuracy issue when enabling cuda graph.

**Review by HaiShaw — COMMENTED** [review]
@sogalin would you please update, we'd like to close it soon.

**Comment by sogalin** [comment]
@amd-bot ci-status
